File: native/core-android/c15t-core/src/test/kotlin/com/c15t/core/tc/VendorListScopeFixturesTest.kt
package com.c15t.core.tc

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertSame
import kotlin.test.assertTrue

/**
 * The publisher's vendor scope, applied to a served list, graded against the shared vectors.
 *
 * The oracle is `narrowGVLToVendors` in `packages/iab/src/tcf/fetch-gvl.ts`: the generator calls
 * that function and writes down what came back, so a difference here is Kotlin and the browser
 * build reading the same document differently, and `native/CONTRACT.md` settles that argument in
 * the web's favour. Three surfaces of this core get the same nine vectors, because the core owns
 * two narrowing paths and only one of them is on the happy path:
 *
 *   * [narrowToVendorIds], the typed prune a decoded list takes on its way into a snapshot, a
 *     bridge payload, and a stored envelope;
 *   * [narrowVendorListElement], the byte-level prune for a served `gvl` that never went through a
 *     decode, which is where the backend stands when it writes its cache;
 *   * [toTcVendorList] on the pruned result, because a vendor that survives a scope is a vendor a
 *     TC String may write a signal for, and one that does not survive must not appear in the
 *     string at all.
 *
 * The claims this file makes are the ones the disclosure promise rests on. The surviving keys and
 * their order come from the fixture rather than from a sort this file chose, so "32 declared
 * partners" is measured rather than assumed; the surviving entries are compared field for field,
 * so a survivor cannot arrive rebuilt; and the records beside `vendors` are compared as served,
 * because a list narrowed to one vendor still has to name what purpose 7 is, and the two version
 * numbers a TC String advertises do not belong to a publisher's scope either.
 */
class VendorListScopeFixturesTest {
	private val json = Json { ignoreUnknownKeys = true }

	private val fixtures: List<VendorListScopeFixture> = VendorListScopeFixtures.load()

	/** The kind these vectors carry, spelled once so the two counts below name the same thing. */
	private val kind: String = VendorListScopeFixtures.KIND

	/** The declared scope in the `Int` keys both narrowing helpers take. `null` stays `null`. */
	private fun scopeOf(fixture: VendorListScopeFixture): List<Int>? =
		fixture.scope?.map { id -> id.toInt() }

	/**
	 * The typed prune's surviving keys, in [Long] for comparison with the fixture.
	 *
	 * The core keys its vendor map by `Int` and the document's keys are numbers written as
	 * strings, so the two meet here rather than in a cast at each call site.
	 */
	private fun typedKeys(fixture: VendorListScopeFixture): List<Long> =
		narrowTyped(fixture).vendors.keys.map { key -> key.toLong() }

	/**
	 * The typed prune. Fails the run if the served document is one this build cannot read, which
	 * would make every other claim here vacuous.
	 */
	private fun narrowTyped(fixture: VendorListScopeFixture): GlobalVendorList {
		val served = GlobalVendorListJson.read(fixture.served)
		assertTrue(served != null, "${fixture.id}: the served document in the fixture is not readable by GlobalVendorListJson.read")
		return served.narrowToVendorIds(scopeOf(fixture))
	}

	/** The byte prune's surviving keys, so the order claim is made in one key space on both paths. */
	private fun byteKeys(fixture: VendorListScopeFixture): List<Long> =
		VendorListScopeFixtures.vendorKeysOf(narrowBytes(fixture))

	private fun narrowBytes(fixture: VendorListScopeFixture): JsonObject {
		val pruned = narrowVendorListElement(fixture.served, scopeOf(fixture))
		assertTrue(pruned is JsonObject, "${fixture.id}: narrowVendorListElement returned ${pruned?.let { it::class.simpleName } ?: "null"} for a served object")
		return pruned
	}

	// -- accounting -----------------------------------------------------------

	/**
	 * Every `vendor-list-scope` entry `index.json` lists is graded by this file.
	 *
	 * Counted from the index rather than from a list of ids written here, so a tenth vector
	 * published by the vectors lane is claimed automatically or reported, and cannot go unrun.
	 */
	@Test
	fun `every vendor-list-scope fixture is claimed`() {
		val claimed = fixtures.map { fixture -> fixture.id }.sorted()
		val listed = VendorListScopeFixtures.ids()
		assertEquals(listed, claimed, "index.json lists ${listed.size} $kind, this file graded ${claimed.size}")
		assertTrue(claimed.isNotEmpty(), "no vendor-list-scope fixtures were loaded, so the counts below are empty rather than green")
		println(
			"VENDOR LIST SCOPE CLAIMS: android claimed=${claimed.size}/${listed.size} unclaimed=0  " +
				"(shapes: ${fixtures.map { fixture -> fixture.shape }.sorted().joinToString(", ")})"
		)
	}

	// -- the surviving keys and their order -----------------------------------

	/**
	 * The surviving `vendors` keys, and the order they survive in, on both narrowing paths.
	 *
	 * Order is not decoration: it is the order the served document put the drawer in, and both
	 * web filters keep it by iterating the document's own keys rather than the order the host
	 * typed the allowlist. Kotlin hands back a `Map` that keeps insertion order on both paths,
	 * so the order the fixture states is comparable directly, on the typed list and on the bytes.
	 */
	@Test
	fun `the surviving vendor keys and their order match the fixture`() {
		for (fixture in fixtures) {
			val expectedOrder = fixture.expectedVendorKeys
			assertEquals(expectedOrder, typedKeys(fixture), "${fixture.id}: the typed prune returned the wrong keys or the wrong order")
			assertEquals(expectedOrder, byteKeys(fixture), "${fixture.id}: the byte prune returned the wrong keys or the wrong order")
			assertEquals(
				expectedOrder,
				VendorListScopeFixtures.vendorKeysOf(fixture.expectedDocument),
				"${fixture.id}: expected.document and expected.vendorKeys disagree with each other"
			)
		}
	}

	/**
	 * A scope that reaches past the last assignment keeps the served order and adds nothing.
	 *
	 * The vector whose scope is 947 ids names every id the framework has ever handed out, and the
	 * answer is exactly the served document: ordering is the document's, and a wider scope buys a
	 * core no extra vendor, which is the same claim the 609-id vector past the query cap makes for
	 * the two served ids that fall outside it.
	 */
	@Test
	fun `a scope wider than the list changes nothing about order`() {
		for (fixture in fixtures.filter { fixture -> fixture.shape == "scope-wider-than-list" || fixture.shape == "scope-above-query-cap" }) {
			val survivors = typedKeys(fixture)
			assertEquals(fixture.servedVendorKeys.filter { id -> fixture.scope!!.contains(id) }, survivors, "${fixture.id}: survivors are the served ids the scope names, in served order")
			assertTrue(survivors.size < fixture.scopeSize, "${fixture.id}: the scope names more vendors than the document carries, so it cannot be satisfied")
		}
	}

	// -- nothing is invented, nothing is renamed ------------------------------

	/**
	 * An id the served document does not carry contributes no entry.
	 *
	 * The unknown-id vector declares three ids, one of which is served, and the answer is one
	 * vendor -- not three. A fabricated entry would reach [toTcVendorList] as a declaration no GVL
	 * revision stands behind, which is contract rule 5 wearing a friendlier face, and the two
	 * numbers here are the ones no company ever published: a made-up six-figure id and one past
	 * the 16 bits a TC String vendor field can even hold.
	 */
	@Test
	fun `a scope id the list does not carry is never added`() {
		val fixture = fixtures.first { candidate -> candidate.shape == "scope-unknown-id" }
		val pruned = narrowBytes(fixture)
		assertEquals(listOf(11L), VendorListScopeFixtures.vendorKeysOf(pruned), "${fixture.id}: only the served id survives")
		val served = GlobalVendorListJson.read(fixture.served)!!
		val narrowed = served.narrowToVendorIds(scopeOf(fixture))
		assertEquals(1, narrowed.vendors.size, "${fixture.id}: the typed prune invented an entry")
		assertTrue(66_001 !in narrowed.vendors.keys, "${fixture.id}: an id past the 16-bit vendor field cannot be disclosed by a TC String anyway")
	}

	/**
	 * Selection follows the map key, and an entry whose body states another id earns nothing by it.
	 *
	 * The vectors plant a pair -- the record keyed 7 states 11, the record keyed 11 states 7 -- so
	 * a filter that consulted bodies would keep the opposite pair for the same scope and every
	 * count in this file would still pass. That is why the claim is about which key survived, which
	 * name arrived with it, and which vendor the encoder is finally told about -- not about how many
	 * entries came back. An entry that could rename itself with its own `id` field could answer for
	 * a vendor the publisher never named.
	 *
	 * The entry's own `id` field is served data and the prune does not touch it, so on this vector
	 * the survivor under key 7 still states 11 in its body. Neither core rewrites it: Kotlin reads
	 * the document as served and takes the identity from the key at [toTcVendorList], where
	 * `@iabtechlabtcf/core`'s `GVL` overwrites `vendor.id` with the key and `mapGvlVendor` in
	 * `dialog-data.ts` does `id: Number(vendorId)`. The graded claim is therefore what the encoder
	 * is told, which is the last place a wrong answer could still reach a subject.
	 */
	@Test
	fun `a vendor is chosen by its key and not by the id inside its body`() {
		val fixture = fixtures.first { candidate -> candidate.shape == "scope-key-not-body-id" }
		val pruned = narrowBytes(fixture)
		assertEquals(listOf(7L), VendorListScopeFixtures.vendorKeysOf(pruned), "${fixture.id}: the record keyed 11 is not in the scope, whatever its body claims")
		val survivor = pruned["vendors"]!!.jsonObject["7"]!!.jsonObject
		assertEquals("Vendor 7", (survivor["name"] as? kotlinx.serialization.json.JsonPrimitive)?.content, "${fixture.id}: the survivor is the entry served under key 7")
		val served = GlobalVendorListJson.read(fixture.served)!!
		val narrowedTyped = served.narrowToVendorIds(scopeOf(fixture))
		val survivorEntry = narrowedTyped.vendors[7]
		assertTrue(survivorEntry != null, "${fixture.id}: key 7 disappeared from the typed prune")
		assertEquals(11L, survivorEntry!!.id, "${fixture.id}: the prune rewrote the served entry's own id field instead of carrying it over")
		val declared = narrowedTyped.toTcVendorList()
		assertTrue(declared.vendor(7) != null, "${fixture.id}: the encoder is not told about the vendor the scope named by key")
		assertTrue(declared.vendor(11) == null, "${fixture.id}: an entry's body id bought a signal the publisher never scoped")

		// The mirror: a scope naming the other half of the pair keeps that record and its name,
		// still not the number it claims to be.
		val mirrored = narrowVendorListElement(fixture.served, listOf(11)) as JsonObject
		assertEquals(listOf(11L), VendorListScopeFixtures.vendorKeysOf(mirrored), "${fixture.id}: the pair is symmetric -- scoping 11 keeps the record keyed 11")

		// And across every vector: no survivor was ever able to buy a different name, because the
		// fixture's own transcript says which key states which id, in both documents.
		for (vector in fixtures) {
			val kept = VendorListScopeFixtures.vendorKeysOf(narrowBytes(vector))
			vector.surviving.forEachIndexed { index, claim ->
				assertEquals(claim.key, kept[index], "${vector.id}: survivor number ${index + 1} is not the key the fixture reported")
				val entry = narrowBytes(vector)["vendors"]!!.jsonObject[claim.key.toString()]!!.jsonObject
				val stated = (entry["id"] as? kotlinx.serialization.json.JsonPrimitive)?.content?.toLongOrNull()
				assertEquals(claim.bodyId, stated, "${vector.id}: the entry keyed ${claim.key} states ${stated}, and the fixture says ${claim.bodyId}")
				if (claim.bodyIdDiffersFromKey) {
					assertTrue(stated != claim.key, "${vector.id}: key ${claim.key} was reported as disagreeing with its body, and now it does not")
				}
			}
		}
	}

	// -- what a survivor keeps ------------------------------------------------

	/**
	 * A surviving entry is the served entry: every field, optional included.
	 *
	 * The prune may remove records and nothing else, so the survivor that leans on `dataRetention`
	 * and `overflow` has to arrive with both, and `carriesOptionalFields` is the fixture's own
	 * statement about which survivors those are. The same comparison is then made whole: the
	 * pruned `vendors` record against the fixture's, byte for byte, which is the claim "carried
	 * over" makes and not the weaker one about counts.
	 */
	@Test
	fun `a survivor arrives with the fields it was served with`() {
		for (fixture in fixtures) {
			val prunedVendors = narrowBytes(fixture)["vendors"]!!.jsonObject
			val servedVendors = fixture.served["vendors"]!!.jsonObject
			val expectedVendors = fixture.expectedDocument["vendors"]!!.jsonObject
			assertEquals(expectedVendors, prunedVendors, "${fixture.id}: the byte prune did not return the document the web filter returned")
			for (claim in fixture.surviving) {
				val key = claim.key.toString()
				val servedEntry = servedVendors[key]
				assertTrue(servedEntry != null, "${fixture.id}: ${claim.key} is served in the fixture and not in the document it narrowed")
				assertEquals(claim.carriesOptionalFields, VendorListScopeFixtures.OPTIONAL_VENDOR_FIELDS.any { field -> servedEntry.jsonObject.containsKey(field) }, "${fixture.id}: ${claim.key} optional-field claim does not match the document")
				if (claim.carriesOptionalFields) {
					assertTrue(
						VendorListScopeFixtures.OPTIONAL_VENDOR_FIELDS.any { field -> prunedVendors[key]!!.jsonObject.containsKey(field) },
						"${fixture.id}: ${claim.key} lost its optional fields on the way through the prune"
					)
				}
			}
		}
	}

	/**
	 * Everything beside `vendors` comes back exactly as served, version numbers included.
	 *
	 * `purposes`, `specialPurposes`, `features`, `specialFeatures`, `stacks`, `dataCategories` and
	 * `lastUpdated` describe the framework rather than the audience, and `processPurposes` in
	 * `packages/iab/src/headless/dialog-data.ts` walks the served `purposes` record to draw a
	 * dialog: pruning an unreferenced purpose for tidiness changes what that dialog claims
	 * independently of the scope anybody configured. The two versions are the expensive pair --
	 * `toTcVendorList` reads them straight into a TC String's VendorListVersion and
	 * TcfPolicyVersion -- so the versions vector numbers them somewhere nobody's default lives,
	 * and the fixture proves a scope of two ids moved neither.
	 */
	@Test
	fun `the records beside vendors are served data and stay served`() {
		for (fixture in fixtures) {
			val pruned = narrowBytes(fixture)
			for (key in VendorListScopeFixtures.SIBLING_KEYS) {
				assertEquals(fixture.served[key], pruned[key], "${fixture.id}: ${key} came back changed, and no scope was asked to touch it")
			}
			assertEquals(
				fixture.served["vendorListVersion"],
				fixture.expectedDocument["vendorListVersion"],
				"${fixture.id}: the expected document did not carry the list version forward"
			)
			assertEquals(
				fixture.served["tcfPolicyVersion"],
				fixture.expectedDocument["tcfPolicyVersion"],
				"${fixture.id}: the expected document did not carry the policy version forward"
			)
		}
	}

	/**
	 * The versions vector keeps numbers nobody defaults to.
	 *
	 * A core that rebuilt a narrowed list from its own constants would come back with the
	 * industry's current pair rather than 431 under policy 4, and the TC String it then encoded
	 * would advertise a list revision and a policy revision that nothing on the wire ever served.
	 */
	@Test
	fun `a narrow scope does not move the two version numbers`() {
		val fixture = fixtures.first { candidate -> candidate.shape == "scope-versions-preserved" }
		val narrowed = narrowTyped(fixture)
		assertEquals(listOf(43L, 62L), typedKeys(fixture), "${fixture.id}: the two declared ids, in served order")
		assertEquals(431L, narrowed.vendorListVersion, "${fixture.id}: vendorListVersion is the served document's, not a constant")
		assertEquals(4L, narrowed.tcfPolicyVersion, "${fixture.id}: tcfPolicyVersion is the served document's, not a constant")
		assertEquals(2, narrowed.vendors.size, "${fixture.id}: two declared ids, two survivors")
	}

	// -- no scope declared ----------------------------------------------------

	/**
	 * An empty scope and an absent scope both mean no scope, and neither empties the drawer.
	 *
	 * This is the fail-open half of the pair, and it is the mistake with the worst shape: a core
	 * that reads `[]` as "show nobody" leaves a publisher who never scoped anything with a
	 * preference centre with no vendors in it, under a consent the subject can still give. Web
	 * reads it that way twice over -- `narrowGVLToVendors` hands the list back for an empty array,
	 * and `gvlRequestUrl` does not append the parameter at all -- and both halves of this core do
	 * the same, on the typed path identity is kept, which is what [assertSame] proves.
	 */
	@Test
	fun `no scope declared leaves the whole list alone`() {
		val empty = fixtures.first { candidate -> candidate.shape == "scope-empty" }
		val absent = fixtures.first { candidate -> candidate.shape == "scope-absent" }
		for (fixture in listOf(empty, absent)) {
			assertTrue(fixture.unchanged, "${fixture.id}: the fixture claims a rebuilt document for a vector that declared no scope")
			val served = GlobalVendorListJson.read(fixture.served)!!
			assertSame(served, served.narrowToVendorIds(scopeOf(fixture)), "${fixture.id}: the typed prune copied a list that declared no scope")
			val untouched = narrowVendorListElement(fixture.served, scopeOf(fixture)) as JsonObject
			assertSame(fixture.served, untouched, "${fixture.id}: the byte prune rebuilt bytes that declared no scope")
			assertEquals(fixture.servedVendorKeys, typedKeys(fixture), "${fixture.id}: every served vendor has to survive")
			assertEquals(fixture.served, fixture.expectedDocument, "${fixture.id}: expected.document is not the document it started from")
		}
	}

	/**
	 * A pruned list still answers the two questions the TC String encoder asks of it.
	 *
	 * [toTcVendorList] is where a surviving vendor becomes a signal a subject can be asked about,
	 * so the pruned list has to still be a list it can build: the two versions carried through,
	 * the surviving vendors present, and nothing from the pruned half in reach. Out of a vector
	 * that declared 32 partners, this is the number the encoder would be able to name.
	 */
	@Test
	fun `the pruned list is still a list the encoder can read`() {
		for (fixture in fixtures) {
			val narrowed = narrowTyped(fixture)
			val tc = narrowed.toTcVendorList()
			assertEquals(narrowed.vendors.size, tc.vendors.size, "${fixture.id}: the encoder dropped or invented a vendor on the way in")
			assertEquals(narrowed.vendorListVersion, tc.vendorListVersion.toLong(), "${fixture.id}: the encoder read the list version off something other than the narrowed list")
			assertEquals(narrowed.tcfPolicyVersion, tc.tcfPolicyVersion.toLong(), "${fixture.id}: the encoder read the policy version off something other than the narrowed list")
		}
	}
}

File: packages/react-native/scripts/__tests__/vendor-list-scope-fixtures.test.ts
/**
 * Proves the committed `vendor-list-scope` fixtures say what they claim.
 *
 * Every test here reads the checked-in JSON and drives `narrowGVLToVendors` again from
 * `fixture.input` alone, the way `tc-string-fixtures.test.ts` drives the reference codec.
 * The generator calls that function too, so a second run from the published side is the
 * assertion that matters: a vector whose `expected` survives it is the web's answer written
 * down, not a transcription of whatever the builder produced on the day it was generated.
 * It is also the only reproduction either native core can manage -- Swift and Kotlin get
 * these bytes and nothing else -- so a fixture that only its own generator could reproduce
 * would read green here and be graded wrong on a device.
 *
 * The SHA-256 in `index.json` is checked against the bytes rather than against a
 * re-serialization, because both native loaders hash before they parse for the same reason:
 * an edited file still parses, and a reader that parsed first would grade the core against
 * somebody else's keystrokes.
 */

import { createHash } from 'node:crypto';
import { readFileSync } from 'node:fs';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

import { describe, expect, test } from 'vitest';

import { narrowGVLToVendors } from '../../../iab/src/tcf/fetch-gvl';
import type { GlobalVendorList } from '../../../iab/src/tcf/iab-tcf-types';
import {
	VENDOR_LIST_SCOPE_FIXTURE_IDS,
	VENDOR_LIST_SCOPE_ORACLE,
	VENDOR_LIST_SCOPE_SHAPES,
} from '../vendor-list-scope-fixtures';

const HERE = dirname(fileURLToPath(import.meta.url));
const FIXTURE_DIR = resolve(HERE, '../../../../native/protocol');
const INDEX_FILE = 'index.json';
const KIND = 'vendor-list-scope';

/** The optional halves of `gvlVendorSchema`, which a rebuilt entry is likeliest to lose. */
const OPTIONAL_VENDOR_FIELDS = [
	'dataCategories',
	'dataRetention',
	'deletedDate',
	'deviceStorageDisclosureUrl',
	'overflow',
] as const;

/** Everything beside `vendors`, plus the two versions a TC String advertises. */
const FRAMEWORK_FIELDS = [
	'dataCategories',
	'features',
	'gvlSpecificationVersion',
	'lastUpdated',
	'purposes',
	'specialFeatures',
	'specialPurposes',
	'stacks',
	'vendorListVersion',
	'tcfPolicyVersion',
] as const;

/** The vendor list as published: id-keyed records whose keys are strings on the wire. */
type VendorRecords = Record<string, Record<string, unknown> | undefined>;

interface IndexEntry {
	bytes: number;
	file: string;
	id: string;
	kind: string;
	protocolVersion: number;
	sha256: string;
}

interface PublishedFixture {
	description: string;
	expected: {
		document: GlobalVendorList;
		unchanged: boolean;
		vendorKeys: string[];
		vendors: {
			bodyId: number | null;
			bodyIdDiffersFromKey: boolean;
			carriesOptionalFields: boolean;
			key: number;
			name: string | null;
		}[];
	};
	id: string;
	input: {
		document: GlobalVendorList;
		scopeSize: number;
		vendorIds: number[] | null;
	};
	kind: typeof KIND;
	notes: string[];
	oracle: typeof VENDOR_LIST_SCOPE_ORACLE;
	protocolVersion: number;
	shape: (typeof VENDOR_LIST_SCOPE_SHAPES)[number];
}

const recordsOf = function recordsOf(
	document: GlobalVendorList
): VendorRecords {
	return document.vendors as unknown as VendorRecords;
};

const entries = (
	JSON.parse(readFileSync(resolve(FIXTURE_DIR, INDEX_FILE), 'utf8')) as {
		fixtures: IndexEntry[];
	}
).fixtures.filter((entry) => entry.kind === KIND);

const fixtures = entries.map((entry) => {
	return JSON.parse(
		readFileSync(resolve(FIXTURE_DIR, entry.file), 'utf8')
	) as PublishedFixture;
});

const vectorFor = function vectorFor(
	shape: PublishedFixture['shape']
): PublishedFixture {
	const found = fixtures.find((fixture) => fixture.shape === shape);
	if (!found) {
		throw new Error(
			`no committed ${KIND} vector has shape ${shape}; the case set lost one`
		);
	}
	return found;
};

/** A survivor transcript rebuilt from the document it is claimed about. */
const survivingFrom = function survivingFrom(
	document: GlobalVendorList,
	vendorKeys: readonly string[]
) {
	const records = recordsOf(document);
	return vendorKeys.map((key) => {
		const entry = records[key];
		return {
			bodyId: (entry?.id as number | undefined) ?? null,
			bodyIdDiffersFromKey: entry === undefined || entry.id !== Number(key),
			carriesOptionalFields: OPTIONAL_VENDOR_FIELDS.some(
				(field) => entry?.[field] !== undefined
			),
			key: Number(key),
			name: (entry?.name as string | undefined) ?? null,
		};
	});
};

describe(`${KIND} fixtures`, () => {
	test('the index pins every vector, and the bytes on disk are the pinned ones', () => {
		expect(entries.map((entry) => entry.id)).toEqual([
			...VENDOR_LIST_SCOPE_FIXTURE_IDS,
		]);
		for (const entry of entries) {
			const bytes = readFileSync(resolve(FIXTURE_DIR, entry.file));
			expect(bytes.byteLength).toBe(entry.bytes);
			expect(createHash('sha256').update(bytes).digest('hex')).toBe(
				entry.sha256
			);
			expect(entry.file).toBe(`${entry.id}.json`);
			expect(entry.id.startsWith(`${KIND}-`)).toBe(true);
			expect(entry.protocolVersion).toBe(1);
		}
	});

	test('every case is published exactly once and describes itself', () => {
		expect(fixtures.length).toBe(VENDOR_LIST_SCOPE_SHAPES.length);
		expect(fixtures.map((fixture) => fixture.shape).sort()).toEqual(
			[...VENDOR_LIST_SCOPE_SHAPES].sort()
		);
		for (const fixture of fixtures) {
			expect(fixture.kind).toBe(KIND);
			// The provenance a runner quotes when it disagrees with a vector.
			expect(fixture.oracle).toStrictEqual(VENDOR_LIST_SCOPE_ORACLE);
			expect(fixture.description.length).toBeGreaterThan(30);
			expect(fixture.notes.length).toBeGreaterThan(0);
			// The three views of the same answer have to agree with each other before any
			// core is graded against them.
			expect(fixture.expected.vendorKeys).toEqual(
				Object.keys(fixture.expected.document.vendors)
			);
			expect(fixture.expected.vendors.map((vendor) => vendor.key)).toEqual(
				fixture.expected.vendorKeys.map((key) => Number(key))
			);
			expect(
				survivingFrom(fixture.expected.document, fixture.expected.vendorKeys)
			).toEqual(fixture.expected.vendors);
			expect(fixture.input.scopeSize).toBe(
				fixture.input.vendorIds?.length ?? 0
			);
		}
	});

	/**
	 * The load-bearing test: every published `input` goes back through the web filter, and
	 * what comes back has to be the published `expected`.
	 *
	 * Object identity is asserted rather than inferred, because `unchanged` claims exactly
	 * that: the filter handed the document back rather than rebuilding a copy that looks
	 * the same. A deep equality alone would let an eager copy satisfy a vector that was
	 * graded as untouched, which is the difference the two no-scope cases exist to pin.
	 */
	test('each expectation is the filter run again on the published input', () => {
		for (const fixture of fixtures) {
			const returned = narrowGVLToVendors(
				fixture.input.document,
				fixture.input.vendorIds ?? []
			);
			expect(returned === fixture.input.document).toBe(
				fixture.expected.unchanged
			);
			expect(returned).toStrictEqual(fixture.expected.document);
			expect(Object.keys(returned.vendors)).toEqual(
				fixture.expected.vendorKeys
			);
		}
	});

	/**
	 * The nine cases, as outcomes. Every count in the set above would still pass with a
	 * filter that fabricated an entry, emptied a drawer nobody scoped, or read a body id, so
	 * each of these names a mistake the counts cannot see.
	 */
	test('each case lands where the disclosure promise says it must', () => {
		// A publisher that declared 32 partners keeps 32, and not one of them is invented.
		const declared = vectorFor('scope-32-vendors');
		expect(declared.input.vendorIds).toHaveLength(32);
		expect(declared.expected.vendorKeys).toHaveLength(32);
		expect(new Set(declared.expected.vendorKeys).size).toBe(32);

		// An empty scope and an absent one both mean nobody scoped anything, so the served
		// list travels whole. An emptied drawer is a refusal the publisher never configured.
		const servedKeys = Object.keys(
			vectorFor('scope-empty').input.document.vendors
		);
		expect(servedKeys).toHaveLength(54);
		// Both documents cross a JSON boundary on their way into this test, so "handed back
		// untouched" is not observable between them -- the re-run above is where `unchanged`
		// gets its teeth. What these two files can show is that the whole document travels.
		for (const shape of ['scope-absent', 'scope-empty'] as const) {
			const fixture = vectorFor(shape);
			expect(fixture.expected.unchanged).toBe(true);
			expect(fixture.expected.document).toStrictEqual(fixture.input.document);
			expect(fixture.expected.vendorKeys).toEqual(servedKeys);
		}
		// Absent is a JSON null and not a missing key: a core has to tell "no scope
		// configured" from "a scope that is empty" even though both answer with the whole
		// list, and `nil` / `null` / `Nothing?` is how both cores receive it.
		expect(vectorFor('scope-absent').input.vendorIds).toBeNull();
		expect(vectorFor('scope-empty').input.vendorIds).toEqual([]);

		// Two of the three ids name nobody. The answer is one vendor out of three, which no
		// count of a declared scope can show -- and an invented entry would put a name on the
		// drawer that no GVL revision stands behind.
		const unknown = vectorFor('scope-unknown-id');
		expect(unknown.input.vendorIds).toEqual([11, 1_000_483, 66_001]);
		expect(unknown.expected.vendorKeys).toEqual(['11']);
		for (const unserved of ['1000483', '66001']) {
			expect(recordsOf(unknown.input.document)[unserved]).toBeUndefined();
			expect(recordsOf(unknown.expected.document)[unserved]).toBeUndefined();
		}

		// Past the last id the framework assigned, a wide scope lands on the served list and
		// adds nothing. It is still a rebuilt document rather than the one handed in, which is
		// why the promise here is "every served vendor survives" and not "untouched".
		const wider = vectorFor('scope-wider-than-list');
		expect(wider.input.scopeSize).toBeGreaterThan(servedKeys.length);
		expect(wider.expected.unchanged).toBe(false);
		expect(wider.expected.vendorKeys).toEqual(servedKeys);

		// Above the 500-id cap the web will put in a request line, the scope cannot travel as a
		// `vendorIds` parameter, so this is the case where the document arrives whole and every
		// caller prunes it itself. The scope names 1..609 and reaches past two served ids, which
		// are dropped exactly like every other id outside it.
		const cap = vectorFor('scope-above-query-cap');
		const capReach = Math.max(...(cap.input.vendorIds ?? []));
		expect(cap.input.scopeSize).toBeGreaterThan(500);
		expect(cap.input.scopeSize).toBe(capReach);
		expect(cap.expected.vendorKeys).toEqual(
			servedKeys.filter((key) => Number(key) <= capReach)
		);
		expect(cap.expected.vendorKeys).not.toContain('755');
		expect(cap.expected.vendorKeys).not.toContain('897');
		expect(cap.expected.vendorKeys).toContain('353');

		// The served pair keyed 7 / 11 states 11 / 7 in its bodies, so a filter that consulted
		// bodies would keep the opposite entry for the same scope and still hand back exactly
		// one vendor. Selection follows the key: key 7 survives, with the served body intact.
		const keyed = vectorFor('scope-key-not-body-id');
		expect(keyed.input.vendorIds).toEqual([7]);
		expect(keyed.expected.vendorKeys).toEqual(['7']);
		expect(recordsOf(keyed.input.document)['7']?.id).toBe(11);
		expect(recordsOf(keyed.input.document)['11']?.id).toBe(7);
		expect(recordsOf(keyed.expected.document)['7']).toStrictEqual(
			recordsOf(keyed.input.document)['7']
		);
		expect(keyed.expected.vendors[0]?.bodyIdDiffersFromKey).toBe(true);
		expect(keyed.expected.document.vendors['11']).toBeUndefined();

		// Survivors keep served order, which is the order the document was pruned in and not
		// the order a host typed the scope. The ids here arrive out of order on purpose.
		const ordered = vectorFor('scope-siblings-preserved');
		expect(ordered.input.vendorIds).toEqual([4, 8, 45, 62, 897]);
		expect(ordered.expected.vendorKeys).toEqual(['4', '8', '45', '62', '897']);

		// The TC String a device encodes advertises both version numbers, so a scope of two
		// ids on a list numbered away from every default has to bring both back.
		const versions = vectorFor('scope-versions-preserved');
		expect(versions.input.document.vendorListVersion).toBe(431);
		expect(versions.input.document.tcfPolicyVersion).toBe(4);
		expect(versions.expected.document.vendorListVersion).toBe(431);
		expect(versions.expected.document.tcfPolicyVersion).toBe(4);
		expect(versions.expected.vendorKeys).toEqual(['43', '62']);
		// Only this vector is renumbered, or the two numbers above would say as much about a
		// builder's defaults as about a served document.
		expect(
			fixtures
				.filter((fixture) => fixture !== versions)
				.every(
					(fixture) =>
						fixture.input.document.vendorListVersion === 177 &&
						fixture.input.document.tcfPolicyVersion === 5
				)
		).toBe(true);
	});

	test('nothing is invented, and every survivor is carried rather than rebuilt', () => {
		for (const fixture of fixtures) {
			const served = recordsOf(fixture.input.document);
			const expected = recordsOf(fixture.expected.document);
			expect(Object.keys(served).length).toBeGreaterThan(0);
			for (const [key, entry] of Object.entries(expected)) {
				// A key with no served record has no entry to survive, so a survivor that
				// reads as undefined here is a fabricated vendor.
				expect(served[key], key).toBeDefined();
				// "Carried over" and not "rebuilt": the entry that survives is the served
				// record, optional fields and all.
				expect(entry, key).toStrictEqual(served[key]);
			}
			// A prune removes records and nothing else, so a key outside the survivor set has
			// to be absent rather than present under repaired bytes.
			const survivors = new Set(fixture.expected.vendorKeys);
			for (const key of Object.keys(served)) {
				if (!survivors.has(key)) {
					expect(expected[key], key).toBeUndefined();
				}
			}
		}
	});

	test('the records beside vendors come back verbatim', () => {
		for (const fixture of fixtures) {
			const served = fixture.input.document as unknown as Record<
				string,
				unknown
			>;
			const expected = fixture.expected.document as unknown as Record<
				string,
				unknown
			>;
			// The framework half has to be populated, or "identical" would be comparing
			// nothing and a pruned `purposes` record would still read clean.
			expect(Object.keys(served.purposes as object).length).toBeGreaterThan(0);
			for (const field of FRAMEWORK_FIELDS) {
				expect(expected[field], field).toStrictEqual(served[field]);
			}
		}
	});
});
